# Indexing and Selecting data


---

## Learning Objectives 


- Select data by position using `.isel()` with values or slices
- Select data by coordinate label/value using `.sel()` with values or slices
- Use nearest-neighbor lookups with `.sel()`
- Use `interp()` to interpolate by coordinate labels

## Prerequisites


| Concepts | Importance | Notes |
| --- | --- | --- |
| [Understanding of xarray core data structures](./01-xarray-fundamentals.ipynb) | Necessary | |
| [Basic familiarity with NumPy indexing](https://numpy.org/doc/stable/reference/arrays.indexing.html) | Helpful | |
| [Basic familiarity with Pandas indexing](https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html) | Helpful | |

- **Time to learn**: *15-20 minutes*



---

## Imports


In [1]:
import xarray as xr

In [2]:
ds = xr.open_dataset(
    "/home/rdevinen/Documents/GitHub/heatGUIde_palm/static_driver/files/riz_2_static", engine="netcdf4")

ds

<xarray.Dataset>
Dimensions:            (nbuilding_pars: 150, nsurface_fraction: 3, x: 64, y: 72, z: 101, zlad: 6, zsoil: 8)
Coordinates:
  * x                  (x) float32 0.5 1.5 2.5 3.5 4.5 ... 60.5 61.5 62.5 63.5
  * y                  (y) float32 0.5 1.5 2.5 3.5 4.5 ... 68.5 69.5 70.5 71.5
  * z                  (z) float32 0.0 0.5 1.5 2.5 3.5 ... 96.5 97.5 98.5 99.5
  * nbuilding_pars     (nbuilding_pars) int32 0 1 2 3 4 ... 145 146 147 148 149
  * zlad               (zlad) float32 0.0 0.5 1.5 2.5 3.5 4.5
  * zsoil              (zsoil) float32 0.005 0.02 0.05 0.1 0.2 0.4 0.8 2.0
  * nsurface_fraction  (nsurface_fraction) int32 0 1 2
Data variables:
    building_id        (y, x) float64 ...
    buildings_2d       (y, x) float32 ...
    buildings_3d       (z, y, x) float32 ...
    zt                 (y, x) float32 ...
    building_type      (y, x) float32 ...
    pavement_type      (y, x) float32 ...
    soil_type          (y, x) float32 ...
    street_type        (y, x) float32 ...
    surface_fraction   (nsurface_fraction, y, x) float32 ...
    vegetation_type    (y, x) float32 ...
    water_type         (y, x) float32 ...
    lad                (zlad, y, x) float32 ...
    building_pars      (nbuilding_pars, y, x) float32 ...
Attributes:
    title:           RIZ static driver with pixel detailing
    author:          devineni
    institution:     Institut für nachhaltige Energiesysteme,Hochschule Offen...
    comment:         Generic crossing example
    creation_date:   22-04-25 15:19:30 +00
    history:         
    keywords:        example, PALM-4U
    license:         
    palm_version:    
    references:      
    source:          PALM trunk
    version:         1
    Conventions:     CF-1.7
    origin_lat:      48.459638
    origin_lon:      7.941967
    origin_time:     2022-02-22 10:00:00 +00
    origin_x:        421779.645
    origin_y:        5367929.731
    origin_z:        0.0
    rotation_angle:  0.0

In [3]:
ds1 = xr.open_dataset(
    "/home/rdevinen/Downloads/ER_Platz_small_static", engine="netcdf4")

ds1

<xarray.Dataset>
Dimensions:            (nbuilding_pars: 47, nsurface_fraction: 3, nvegetation_pars: 12, x: 128, y: 128, z: 122, zlad: 44)
Coordinates:
    E_UTM              (y, x) float32 ...
    N_UTM              (y, x) float32 ...
    lat                (y, x) float32 ...
    lon                (y, x) float32 ...
  * nbuilding_pars     (nbuilding_pars) int32 0 1 2 3 4 5 ... 41 42 43 44 45 46
  * nsurface_fraction  (nsurface_fraction) int32 0 1 2
  * nvegetation_pars   (nvegetation_pars) int32 0 1 2 3 4 5 6 7 8 9 10 11
  * x                  (x) float32 856.5 857.5 858.5 859.5 ... 981.5 982.5 983.5
  * y                  (y) float32 856.5 857.5 858.5 859.5 ... 981.5 982.5 983.5
  * z                  (z) float32 0.0 0.5 1.5 2.5 ... 117.5 118.5 119.5 120.5
  * zlad               (zlad) float32 0.0 0.5 1.5 2.5 ... 39.5 40.5 41.5 42.5
Data variables:
    bad                (zlad, y, x) float32 ...
    building_id        (y, x) float64 ...
    building_type      (y, x) float32 ...
    buildings_2d       (y, x) float32 ...
    buildings_3d       (z, y, x) float32 ...
    crs                int32 ...
    lad                (zlad, y, x) float32 ...
    pavement_type      (y, x) float32 ...
    soil_type          (y, x) float32 ...
    street_crossing    (y, x) float32 ...
    street_type        (y, x) float32 ...
    surface_fraction   (nsurface_fraction, y, x) float32 ...
    tree_id            (zlad, y, x) float32 ...
    vegetation_pars    (nvegetation_pars, y, x) float32 ...
    vegetation_type    (y, x) float32 ...
    water_type         (y, x) float32 ...
    zt                 (y, x) float32 ...
    building_pars      (nbuilding_pars, y, x) float32 ...
Attributes:
    Conventions:     CF-1.7
    origin_x:        385196.5
    origin_y:        5818435.0
    origin_time:     
    origin_lat:      52.503944396972656
    origin_lon:      13.308590412139893
    acronym:         LUHimuk
    rotation_angle:  0.0
    author:          Suehring, Matthias, suehring@muk.uni-hannover.de
    campaign:        winter iop 1
    comment:         created with palm_csd
    contact:         Suehring, Matthias, suehring@muk.uni-hannover.de
    data_content:    text
    dependencies:    text
    institution:     Institute of Meteorology and Climatology, Leibniz Univer...
    keywords:        text
    location:        Berlin
    palm_version:    6.0
    references:      text
    site:            Ernst Reuter Platz
    source:          text
    version:         1.0
    origin_z:        10.23086929321289
    history:         Thu Feb 24 10:16:36 2022: ncrename -v building_pars_save...
    NCO:             4.7.2

In [4]:
ds2 = xr.open_dataset(
    "/home/rdevinen/Downloads/floating_overflow/THFc7p1N/INPUT/THFc7p1N_static", engine="netcdf4")

ds2

<xarray.Dataset>
Dimensions:            (nbuilding_pars: 47, nsurface_fraction: 3, nvegetation_pars: 12, x: 424, y: 424, z: 39, zlad: 22)
Coordinates:
  * x                  (x) float32 5.0 15.0 25.0 35.0 ... 4215.0 4225.0 4235.0
  * y                  (y) float32 5.0 15.0 25.0 35.0 ... 4215.0 4225.0 4235.0
    lat                (y, x) float32 ...
    lon                (y, x) float32 ...
    E_UTM              (y, x) float32 ...
    N_UTM              (y, x) float32 ...
  * z                  (z) float32 0.0 1.0 3.0 5.0 7.0 ... 69.0 71.0 73.0 75.0
  * nsurface_fraction  (nsurface_fraction) int32 0 1 2
  * nbuilding_pars     (nbuilding_pars) int32 0 1 2 3 4 5 ... 41 42 43 44 45 46
  * nvegetation_pars   (nvegetation_pars) int32 0 1 2 3 4 5 6 7 8 9 10 11
  * zlad               (zlad) float32 0.0 1.0 3.0 5.0 ... 35.0 37.0 39.0 41.0
Data variables:
    crs                int32 ...
    zt                 (y, x) float32 ...
    buildings_2d       (y, x) float32 ...
    building_id        (y, x) float64 ...
    building_type      (y, x) float32 ...
    buildings_3d       (z, y, x) float32 ...
    surface_fraction   (nsurface_fraction, y, x) float32 ...
    vegetation_type    (y, x) float32 ...
    pavement_type      (y, x) float32 ...
    water_type         (y, x) float32 ...
    soil_type          (y, x) float32 ...
    street_type        (y, x) float32 ...
    street_crossing    (y, x) float32 ...
    building_pars      (nbuilding_pars, y, x) float32 ...
    vegetation_pars    (nvegetation_pars, y, x) float32 ...
    lad                (zlad, y, x) float32 ...
    bad                (zlad, y, x) float32 ...
    tree_id            (zlad, y, x) float64 ...
    tree_type          (zlad, y, x) float32 ...
Attributes:
    Conventions:     CF-1.7
    origin_x:        389841.5
    origin_y:        5812984.0
    origin_time:     
    origin_lat:      52.45591354370117
    origin_lon:      13.378795623779297
    acronym:         TUBklima
    rotation_angle:  0.0
    author:          Achim Holtmann, achim.holtmann@tu-berlin.de
    campaign:        none
    comment:         created with palm_csd
    contact:         Achim Holtmann, achim.holtmann@tu-berlin.de
    data_content:    Static driver data
    dependencies:    none
    institution:     Technische Universität Berlin, Fachgebiet Klimatologie
    keywords:        none
    location:        Berlin
    palm_version:    6.0
    references:      none
    site:            Tempelhofer Feld
    source:          misc
    version:         1.0
    origin_z:        31.64315414428711

## NumPy Positional Indexing

When working with numpy, indexing is done by position (slices/ranges/scalars).

In [5]:
sf = ds1["building_pars"]  # retrieve numpy array
sf[4,:,:]

<xarray.DataArray 'building_pars' (y: 128, x: 128)>
array([[ 2.,  2.,  2., ...,  2.,  2.,  2.],
       [ 2.,  2.,  2., ...,  2.,  2.,  2.],
       [ 2.,  2.,  2., ...,  2.,  2.,  2.],
       ...,
       [nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan],
       [nan, nan, nan, ..., nan, nan, nan]], dtype=float32)
Coordinates:
    E_UTM           (y, x) float32 386053.0 386054.0 ... 386179.0 386180.0
    N_UTM           (y, x) float32 5819291.5 5819291.5 ... 5819418.5 5819418.5
    lat             (y, x) float32 52.511818 52.511818 ... 52.512985 52.512985
    lon             (y, x) float32 13.3209095 13.320925 ... 13.322722 13.322737
    nbuilding_pars  int32 4
  * x               (x) float32 856.5 857.5 858.5 859.5 ... 981.5 982.5 983.5
  * y               (y) float32 856.5 857.5 858.5 859.5 ... 981.5 982.5 983.5
Attributes:
    long_name:     building_pars
    units:         
    res_orig:      1.0
    grid_mapping:  E_UTM N_UTM lon lat

In [6]:
sf.shape, sf.ndim

((47, 128, 128), 3)

Let's extract a timeseries for a single spatical location 


In [7]:
sf[:, 32, 0]

<xarray.DataArray 'building_pars' (nbuilding_pars: 47)>
array([nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
       nan, nan, nan, nan, nan], dtype=float32)
Coordinates:
    E_UTM           float32 386053.0
    N_UTM           float32 5819323.5
    lat             float32 52.512104
    lon             float32 13.320899
  * nbuilding_pars  (nbuilding_pars) int32 0 1 2 3 4 5 6 ... 41 42 43 44 45 46
    x               float32 856.5
    y               float32 888.5
Attributes:
    long_name:     building_pars
    units:         
    res_orig:      1.0
    grid_mapping:  E_UTM N_UTM lon lat

## Different choices for indexing 


Xarray supports two kinds of indexing 

- Positional indexing via `.isel()`: provides primarily integer position based index (from `0` to `length-1` of the axis/dimension
- Label indexing via `.sel()`: provides primarily label based index

Xarray's indexing methods preserves the coordinate labels and associated metadata.



### Selection by position

The `.isel()` method is the primary access method for **purely integer based indexing**. The following are valid inputs:
- An integer e.g. `lat=10`
- A list or array of integers `lon=[10, 20, 39]`
- A slice object with integers e.g. `time=slice(2, 20)`

In [8]:
ds["surface_fraction"].isel()  # the original object i.e. no selection

<xarray.DataArray 'surface_fraction' (nsurface_fraction: 3, y: 72, x: 64)>
array([[[1., 1., ..., 1., 1.],
        [1., 1., ..., 1., 1.],
        ...,
        [1., 1., ..., 1., 1.],
        [1., 1., ..., 1., 1.]],

       [[0., 0., ..., 0., 0.],
        [0., 0., ..., 0., 0.],
        ...,
        [0., 0., ..., 0., 0.],
        [0., 0., ..., 0., 0.]],

       [[0., 0., ..., 0., 0.],
        [0., 0., ..., 0., 0.],
        ...,
        [0., 0., ..., 0., 0.],
        [0., 0., ..., 0., 0.]]], dtype=float32)
Coordinates:
  * x                  (x) float32 0.5 1.5 2.5 3.5 4.5 ... 60.5 61.5 62.5 63.5
  * y                  (y) float32 0.5 1.5 2.5 3.5 4.5 ... 68.5 69.5 70.5 71.5
  * nsurface_fraction  (nsurface_fraction) int32 0 1 2
Attributes:
    long_name:  surface fraction
    units:      1

In [9]:
ds["surface_fraction"].isel(x=32)

<xarray.DataArray 'surface_fraction' (nsurface_fraction: 3, y: 72)>
array([[1., 1., 1., ..., 1., 1., 1.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]], dtype=float32)
Coordinates:
    x                  float32 32.5
  * y                  (y) float32 0.5 1.5 2.5 3.5 4.5 ... 68.5 69.5 70.5 71.5
  * nsurface_fraction  (nsurface_fraction) int32 0 1 2
Attributes:
    long_name:  surface fraction
    units:      1

In [10]:
ds["surface_fraction"].isel(x=32, y=0)

<xarray.DataArray 'surface_fraction' (nsurface_fraction: 3)>
array([1., 0., 0.], dtype=float32)
Coordinates:
    x                  float32 32.5
    y                  float32 0.5
  * nsurface_fraction  (nsurface_fraction) int32 0 1 2
Attributes:
    long_name:  surface fraction
    units:      1

In [11]:
ds["surface_fraction"].isel(x=32, y=slice(10, 35))

<xarray.DataArray 'surface_fraction' (nsurface_fraction: 3, y: 25)>
array([[ 0.,  0.,  0.,  0.,  0.,  0.,  0., nan, nan, nan, nan, nan, nan, nan,
        nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan],
       [ 1.,  1.,  1.,  1.,  1.,  1.,  1., nan, nan, nan, nan, nan, nan, nan,
        nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan],
       [ 0.,  0.,  0.,  0.,  0.,  0.,  0., nan, nan, nan, nan, nan, nan, nan,
        nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan]], dtype=float32)
Coordinates:
    x                  float32 32.5
  * y                  (y) float32 10.5 11.5 12.5 13.5 ... 31.5 32.5 33.5 34.5
  * nsurface_fraction  (nsurface_fraction) int32 0 1 2
Attributes:
    long_name:  surface fraction
    units:      1

### Selection by label 


The `.sel()` method is the primary access method for **purely coordinate label based indexing.**. The following are valid inputs:

- A single coordinate label e.g. `time="2021-03-01"`
- A list or array of coordinate labels `lon=[="2021-01-01", ="2021-03-10", ="2021-03-12"]`
- A slice object with coordinate labels e.g. `time=slice("2021-01-01", "2021-03-01")`.  (Note that contrary to usual Python slices, both the start and the stop are included, when present in the index!)

In [12]:
ds["surface_fraction"].sel(y=12.5, method='nearest')

<xarray.DataArray 'surface_fraction' (nsurface_fraction: 3, x: 64)>
array([[1., 1., 1., 1., 1., 1., 1., 1., 1., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 1., 1., 1., 1., 1., 1.,
        1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
        1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
        1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
        0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]], dtype=float32)
Coordinates:
  * x                  (x) float32 0.5 1.5 2.5 3.5 4.5 ... 60.5 61.5 62.5 63.5
    y                  float32 12.5
  * nsurface_fraction  (nsurface_fraction) int32 0 1 2
Attributes:
    long_name:  surface fraction
    units:      1

### Nearest-neighbor lookups

As shown above, when our coordinate labels are not integers or strings or datetime-like but floating point numbers, `.sel()` may throw a `KeyError`:

`ds.tas.sel(lat=39.5, lon=105.7)` fails because we are trying to use a conditional for an approximate value i.e floating numbers are represented approximately inside the computer, and xarray is unable to locate this exact value. To address this issue, xarray supports `method` and `tolerance` keyword argument. The `method` parameter allows for enabling nearest neighbor (inexact) lookups by use of the methods `'pad', 'backfill' or 'nearest'`: 

In [13]:
ds["surface_fraction"].sel(x=32, y=0, method='nearest')

<xarray.DataArray 'surface_fraction' (nsurface_fraction: 3)>
array([1., 0., 0.], dtype=float32)
Coordinates:
    x                  float32 32.5
    y                  float32 0.5
  * nsurface_fraction  (nsurface_fraction) int32 0 1 2
Attributes:
    long_name:  surface fraction
    units:      1

See the [xarray documentation](https://xarray.pydata.org/en/stable/generated/xarray.DataArray.sel.html) for more on usage of `method` and `tolerance` parameters in `.sel()`. 

<div class="admonition alert alert-info">
    <p class="title" style="font-weight:bold">Tip</p>
Another way to use the nearest neighbor lookup is via slice objects. For e.g.:
</div>

In [14]:
ds["surface_fraction"].sel(x=slice(31, 33), y=slice(0, 1))

<xarray.DataArray 'surface_fraction' (nsurface_fraction: 3, y: 1, x: 2)>
array([[[1., 1.]],

       [[0., 0.]],

       [[0., 0.]]], dtype=float32)
Coordinates:
  * x                  (x) float32 31.5 32.5
  * y                  (y) float32 0.5
  * nsurface_fraction  (nsurface_fraction) int32 0 1 2
Attributes:
    long_name:  surface fraction
    units:      1

Operators can be chained, so multiple operations can be peformed sequentially. For example, to select an area of interest and the first time index

In [15]:
ds["surface_fraction"].isel(nsurface_fraction=0).sel(x=slice(31, 33), y=slice(0, 1))

<xarray.DataArray 'surface_fraction' (y: 1, x: 2)>
array([[1., 1.]], dtype=float32)
Coordinates:
  * x                  (x) float32 31.5 32.5
  * y                  (y) float32 0.5
    nsurface_fraction  int32 0
Attributes:
    long_name:  surface fraction
    units:      1

### Interpolation

If we want to interpolate along coordinates rather than looking up the nearest neighbos, we can use the `.interp()` method. To use `interp()` requires the presence of `scipy` library. 


In [16]:
ds["surface_fraction"].interp(lat=[10, 10.1, 10.2], method='nearest')

ValueError: dimensions {'lat'} do not exist. Expected one or more of Frozen(SortedKeysDict({'x': 64, 'y': 72, 'nsurface_fraction': 3}))

---

## Summary 

- Xarray’s named dimensions and labeled coordinates free the user from having to track positional ordering of dimensions when accessing data
- Xarray provides a variety of methods for subsetting data via `.sel()`, `.isel()`, `.interp()` methods


## Resources and References

- [Xarray Documentation - Indexing and Selecting Data](https://xarray.pydata.org/en/stable/indexing.html)
- [Xarray Documentation - Interpolation](https://xarray.pydata.org/en/stable/user-guide/interpolation.html)


<div class="admonition alert alert-success">
    <p class="title" style="font-weight:bold">Previous: <a href="./01-xarray-fundamentals.ipynb">Xarray Fundamentals</a></p>
    <p class="title" style="font-weight:bold">Next: <a href="./03-data-visualization.ipynb">Data Visualization</a></p>
</div>